src/main/java/com/puppycrawl/tools/checkstyle/xpath/XpathQueryGenerator.java

```diff
@@ -0,0 +1,217 @@
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.xpath;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.TokenUtils;
+
+/**
+ * Generates xpath queries. Xpath queries are generated based on received
+ * {@code DetailAst} element, line number and column number.
+ *
+ * <p>
+ *     Example class
+ * </p>
+ * <pre>
+ * public class Main {
+ *
+ *     public String sayHello(String name) {
+ *         return "Hello, " + name;
+ *     }
+ * }
+ * </pre>
+ *
+ * <p>
+ *     Following expression returns list of queries. Each query is the string representing full
+ *     path to the node inside Xpath tree, whose line number is 3 and column number is 4.
+ * </p>
+ * <pre>
+ *     new XpathQueryGenerator(rootAst, 3, 4).generate();
+ * </pre>
+ *
+ * <p>
+ *     Result list
+ * </p>
+ * <ul>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS
+ *     </li>
+ *     <li>
+ *         /CLASS_DEF[@text='Main']/OBJBLOCK/METHOD_DEF[@text='sayHello']/MODIFIERS/LITERAL_PUBLIC
+ *     </li>
+ * </ul>
+ *
+ * @author Timur Tibeyev.
+ */
+public class XpathQueryGenerator {
+    /** The root ast. */
+    private final DetailAST rootAst;
+    /** The line number. */
+    private final int lineNumber;
+    /** The column number. */
+    private final int columnNumber;
+
+    /**
+     * Creates a new {@code XpathQueryGenerator} instance.
+     *
+     * @param rootAst root ast
+     * @param lineNumber line number
+     * @param columnNumber column number
+     */
+    public XpathQueryGenerator(DetailAST rootAst, int lineNumber, int columnNumber) {
+        this.rootAst = rootAst;
+        this.lineNumber = lineNumber;
+        this.columnNumber = columnNumber;
+    }
+
+    /**
+     * Returns list of xpath queries of nodes, matching line and column number.
+     * This approach uses traversion through DetailAst tree.
```

Comment:
@timurt 
>  DetailAST means detail abstract syntax tree.

It is not necessary, please remove. I meant that DetailAST already contains the word `tree`.